File: android/src/main/kotlin/ch/waio/pro_video_editor/src/features/render/helpers/CompositionBuilder.kt
package ch.waio.pro_video_editor.src.features.render.helpers

import RENDER_TAG
import android.content.Context
import android.media.MediaExtractor
import android.media.MediaFormat
import android.util.Log
import androidx.media3.common.Effect
import androidx.media3.common.audio.AudioProcessor
import androidx.media3.common.util.UnstableApi
import androidx.media3.transformer.Composition
import androidx.media3.transformer.EditedMediaItemSequence
import ch.waio.pro_video_editor.src.features.render.models.RenderConfig
import ch.waio.pro_video_editor.src.features.render.models.VideoClip

/**
 * Main builder class for creating Media3 Compositions from render configurations.
 * 
 * Orchestrates video sequences, custom audio tracks, audio normalization,
 * and sample rate compatibility checking. This class delegates the actual
 * building to specialized builders (VideoSequenceBuilder, AudioSequenceBuilder).
 */
@UnstableApi
class CompositionBuilder(
    private val context: Context,
    private val config: RenderConfig,
    private val onAudioMixProgress: ((Double) -> Unit)? = null
) {
    
    private var videoEffects: List<Effect> = emptyList()
    private var audioEffects: List<AudioProcessor> = emptyList()

    /**
     * Sets the video effects to apply from EffectsProcessor.
     */
    fun setVideoEffects(effects: List<Effect>): CompositionBuilder {
        this.videoEffects = effects
        return this
    }

    /**
     * Sets the audio effects to apply from EffectsProcessor.
     */
    fun setAudioEffects(effects: List<AudioProcessor>): CompositionBuilder {
        this.audioEffects = effects
        return this
    }

    /**
     * Builds the complete composition with video and optional custom audio.
     * 
     * @return Composition ready for Media3 Transformer, or null if no video clips
     */
    fun build(): Composition? {
        if (config.videoClips.isEmpty()) {
            return null
        }

        Log.d(RENDER_TAG, "Creating composition with ${config.videoClips.size} video clips")
        Log.d(RENDER_TAG, "Audio enabled: ${config.enableAudio}")

        val rotationDegrees = (4 - (config.rotateTurns ?: 0)) * 90f

        // Build video sequence
        val videoBuilder = VideoSequenceBuilder(config.videoClips)
            .setVideoEffects(videoEffects)
            .setAudioEffects(audioEffects)
            .setRotation(rotationDegrees)
            .setFlip(config.flipX, config.flipY)
            .setCrop(config.cropWidth, config.cropHeight, config.cropX, config.cropY)
            .setImageLayer(config.imageBytes, config.scaleX, config.scaleY)
            .setEnableAudio(config.enableAudio)
            .setOriginalAudioVolume(config.originalAudioVolume)
            .setGlobalTrim(config.startUs, config.endUs)

        // Detect if audio normalization is needed
        val needsNormalization = videoBuilder.detectAudioNormalizationNeeded()
        videoBuilder.setAudioNormalization(needsNormalization)

        // Check if we need to mix custom audio with original
        val needsAudioMixing = config.customAudioPath != null && 
                               config.customAudioPath.isNotEmpty() &&
                               config.originalAudioVolume != null && 
                               config.originalAudioVolume > 0.0f

        // If mixing, we need to remove audio from video and provide mixed track separately
        val forceRemoveOriginalAudio = needsAudioMixing

        videoBuilder.setForceRemoveAudio(forceRemoveOriginalAudio)

        // Build video sequence (with or without audio based on mixing needs)
        val videoSequence = videoBuilder.build()
        
        // Prepare sequences list
        val sequences = mutableListOf<EditedMediaItemSequence>()
        sequences.add(videoSequence)
        Log.d(RENDER_TAG, "Created video EditedMediaItemSequence with ${config.videoClips.size} items")

        // Log audio mixing status
        if (needsAudioMixing) {
            Log.d(
                RENDER_TAG,
                "✅ Audio mixing ENABLED (original: ${config.originalAudioVolume}x, custom: ${config.customAudioVolume}x)"
            )
            Log.d(RENDER_TAG, "Both original and custom audio tracks will be mixed")
        }

        // Add custom audio sequence if provided - mix it with video audio
        if (config.customAudioPath != null && config.customAudioPath.isNotEmpty()) {
            val totalVideoDuration = videoBuilder.calculateTotalDuration()
            
            // Get first video path for audio extraction
            val videoInputPath = config.videoClips.firstOrNull()?.inputPath
            
            if (videoInputPath != null && needsAudioMixing) {
                // Mix both audios into a single track (preserves video's original audio quality)
                Log.d(RENDER_TAG, "🎵 Mixing video audio with custom audio (preserves original quality)")
                
                val audioMixer = AudioMixer(context)
                val mixedAudioPath = audioMixer.mixAudio(
                    videoPath = videoInputPath,
                    customAudioPath = config.customAudioPath,
                    videoVolume = config.originalAudioVolume ?: 1f,
                    customVolume = config.customAudioVolume ?: 1f,
                    targetDuration = totalVideoDuration,
                    onProgress = onAudioMixProgress
                )
                
                Log.d(RENDER_TAG, "Mixed audio created: $mixedAudioPath")
                
                // Add mixed audio sequence (video already has audio removed)
                val audioSequence = AudioSequenceBuilder(mixedAudioPath, totalVideoDuration)
                    .setVolume(1.0f) // Already mixed with correct volumes
                    .setNormalization(needsNormalization)
                    .build()

                if (audioSequence != null) {
                    sequences.add(audioSequence)
                    Log.d(RENDER_TAG, "Mixed audio sequence added")
                }
            } else if (!needsAudioMixing) {
                // Only custom audio, no mixing needed
                Log.d(RENDER_TAG, "Only custom audio (no video audio mixing)")
                val audioSequence = AudioSequenceBuilder(config.customAudioPath, totalVideoDuration)
                    .setVolume(config.customAudioVolume ?: 1.0f)
                    .setNormalization(needsNormalization)
                    .build()

                if (audioSequence != null) {
                    sequences.add(audioSequence)
                    Log.d(RENDER_TAG, "Custom audio sequence added")
                }
            }
        }

        // Build final composition
        val composition = Composition.Builder(sequences).build()
        Log.d(RENDER_TAG, "Composition created successfully with ${sequences.size} sequences")

        return composition
    }
}

File: macos/Classes/src/features/render/helpers/CompositionBuilder.swift
import AVFoundation
import Foundation

/// Main builder class for creating video compositions from render configurations.
///
/// Orchestrates video sequences, custom audio tracks, and audio mixing.
/// This class delegates the actual work to specialized builders
/// (VideoSequenceBuilder, AudioSequenceBuilder) following the Builder pattern.
internal class CompositionBuilder {
    
    private let videoClips: [VideoClip]
    private let videoEffects: VideoCompositorConfig
    private var enableAudio: Bool = true
    private var customAudioPath: String?
    private var originalAudioVolume: Float = 1.0
    private var customAudioVolume: Float = 1.0
    
    /// Initializes builder with configuration.
    ///
    /// - Parameters:
    ///   - videoClips: Array of video clips to process
    ///   - videoEffects: Video effect configuration
    init(videoClips: [VideoClip], videoEffects: VideoCompositorConfig) {
        self.videoClips = videoClips
        self.videoEffects = videoEffects
    }
    
    /// Enables or disables audio.
    ///
    /// - Parameter enabled: If true, includes original audio from video clips
    /// - Returns: Self for chaining
    func setEnableAudio(_ enabled: Bool) -> CompositionBuilder {
        self.enableAudio = enabled
        return self
    }
    
    /// Sets custom audio path.
    ///
    /// - Parameter path: Path to custom audio file
    /// - Returns: Self for chaining
    func setCustomAudioPath(_ path: String?) -> CompositionBuilder {
        self.customAudioPath = path
        return self
    }
    
    /// Sets volume for original video audio.
    ///
    /// - Parameter volume: Volume multiplier (0.0 to 1.0+)
    /// - Returns: Self for chaining
    func setOriginalAudioVolume(_ volume: Float?) -> CompositionBuilder {
        self.originalAudioVolume = volume ?? 1.0
        return self
    }
    
    /// Sets volume for custom audio.
    ///
    /// - Parameter volume: Volume multiplier (0.0 to 1.0+)
    /// - Returns: Self for chaining
    func setCustomAudioVolume(_ volume: Float?) -> CompositionBuilder {
        self.customAudioVolume = volume ?? 1.0
        return self
    }
    
    /// Builds the complete composition.
    ///
    /// - Returns: Tuple containing composition, video composition, render size, and audio mix
    /// - Throws: Error if composition creation fails
    func build() async throws -> (AVMutableComposition, AVMutableVideoComposition, CGSize, AVAudioMix?) {
        guard !videoClips.isEmpty else {
            throw NSError(
                domain: "CompositionBuilder",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Video clips cannot be empty"]
            )
        }
        
        print("🎬 Creating composition with \(videoClips.count) video clips")
        print("🔊 Audio enabled: \(enableAudio)")
        
        let composition = AVMutableComposition()
        
        // Build video sequence
        let videoBuilder = VideoSequenceBuilder(videoClips: videoClips)
            .setEnableAudio(enableAudio)
            .setOriginalAudioVolume(originalAudioVolume)
        
        // Check if we need to mix custom audio with original
        let needsAudioMixing = customAudioPath != nil &&
                               !(customAudioPath?.isEmpty ?? true) &&
                               originalAudioVolume > 0.0
        
        // Check sample rate compatibility if mixing audio
        var forceRemoveOriginalAudio = false
        if needsAudioMixing, let customPath = customAudioPath {
            let audioBuilder = AudioSequenceBuilder(
                audioPath: customPath,
                targetDuration: await videoBuilder.calculateTotalDuration()
            )
            let isCompatible = await audioBuilder.checkSampleRateCompatibility(videoClips: videoClips)
            forceRemoveOriginalAudio = !isCompatible
            
            if forceRemoveOriginalAudio {
                print("❌ Audio mixing DISABLED - sample rate mismatch detected")
                print("❌ Only custom audio will be used (original audio removed)")
            } else {
                print("✅ Audio mixing ENABLED (original: \(originalAudioVolume)x, custom: \(customAudioVolume)x)")
                print("✅ Sample rates are compatible - both audio tracks will be mixed")
            }
        }
        
        // Build video sequence (may force remove audio if incompatible)
        if forceRemoveOriginalAudio {
            _ = videoBuilder.setEnableAudio(false)
        }
        
        let videoResult = try await videoBuilder.build(in: composition)
        
        // Add custom audio track if provided
        var customAudioTrack: AVMutableCompositionTrack?
        if let customPath = customAudioPath, !customPath.isEmpty {
            print("🎵 Adding custom audio track: \(customPath)")
            let audioBuilder = AudioSequenceBuilder(
                audioPath: customPath,
                targetDuration: videoResult.totalDuration
            ).setVolume(customAudioVolume)
            
            customAudioTrack = try await audioBuilder.build(in: composition)
        }
        
        // Create audio mix with volume parameters
        var audioMix: AVAudioMix?
        if enableAudio && !forceRemoveOriginalAudio && (originalAudioVolume != 1.0 || customAudioVolume != 1.0) {
            audioMix = createAudioMix(
                originalTracks: videoResult.audioTracks,
                customTrack: customAudioTrack,
                originalVolume: originalAudioVolume,
                customVolume: customAudioVolume
            )
        }
        
        // Create video composition
        let videoComposition = AVMutableVideoComposition()
        videoComposition.frameDuration = CMTime(
            value: 1,
            timescale: Int32(max(30, videoResult.frameRate))
        )
        videoComposition.renderSize = videoResult.renderSize
        
        // Create single instruction for the entire composition track
        let instruction = AVMutableVideoCompositionInstruction()
        instruction.timeRange = CMTimeRange(start: .zero, duration: videoResult.totalDuration)
        instruction.backgroundColor = CGColor(red: 0, green: 0, blue: 0, alpha: 1)
        
        // Create layer instruction for the composition video track
        let layerInstruction = AVMutableVideoCompositionLayerInstruction(
            assetTrack: videoResult.videoTrack
        )
        layerInstruction.setTransform(videoResult.transform, at: .zero)
        instruction.layerInstructions = [layerInstruction]
        
        videoComposition.instructions = [instruction]
        
        print("✅ Composition created successfully with \(videoClips.count) clips")
        
        return (composition, videoComposition, videoResult.renderSize, audioMix)
    }
    
    /// Creates audio mix with volume parameters.
    private func createAudioMix(
        originalTracks: [AVMutableCompositionTrack],
        customTrack: AVMutableCompositionTrack?,
        originalVolume: Float,
        customVolume: Float
    ) -> AVAudioMix {
        var audioMixInputParameters: [AVMutableAudioMixInputParameters] = []
        
        // Apply volume to original audio tracks
        for track in originalTracks {
            let inputParameters = AVMutableAudioMixInputParameters(track: track)
            inputParameters.setVolume(originalVolume, at: .zero)
            audioMixInputParameters.append(inputParameters)
            print("🔊 Applied volume \(originalVolume) to original audio track")
        }
        
        // Apply volume to custom audio track
        if let customTrack = customTrack {
            let inputParameters = AVMutableAudioMixInputParameters(track: customTrack)
            inputParameters.setVolume(customVolume, at: .zero)
            audioMixInputParameters.append(inputParameters)
            print("🔊 Applied volume \(customVolume) to custom audio track")
        }
        
        let audioMix = AVMutableAudioMix()
        audioMix.inputParameters = audioMixInputParameters
        
        return audioMix
    }
}
